service/graph.go

```diff
@@ -356,3 +356,43 @@ func (p *pipelineNodes) exporterIDs() []string {
 func (p *pipelineNodes) mutatesData() bool {
 	return p.capabilitiesNode.getConsumer().Capabilities().MutatesData
 }
+
+func cycleErr(err error, cycles [][]graph.Node) error {
+	var topoErr topo.Unorderable
+	if !errors.As(err, &topoErr) || len(cycles) == 0 || len(cycles[0]) == 0 {
+		return err
+	}
+
+	// There may be multiple cycles, but report only the first one.
+	cycle := cycles[0]
+
+	// The last node is a duplicate of the first node.
+	// Remove it because we may start from a different node.
+	cycle = cycle[:len(cycle)-1]
+
+	// A cycle always contains a connector. For the sake of consistent
+	// error messages report the cycle starting from a connector.
+	for i := 0; i < len(cycle); i++ {
+		if _, ok := cycle[i].(*connectorNode); ok {
+			cycle = append(cycle[i:], cycle[:i]...)
+			break
+		}
+	}
+
+	// Repeat the first node at the end to clarify the cycle
+	cycle = append(cycle, cycle[0])
+
+	// Build the error message
+	componentDetails := make([]string, 0, len(cycle))
+	for _, node := range cycle {
+		switch n := node.(type) {
+		case *processorNode:
+			componentDetails = append(componentDetails, fmt.Sprintf("processor %q in pipeline %q", n.componentID, n.pipelineID))
+		case *connectorNode:
+			componentDetails = append(componentDetails, fmt.Sprintf("connector %q (%s to %s)", n.componentID, n.exprPipelineType, n.rcvrPipelineType))
+		default:
+			continue // skip capabilities/fanout nodes
+		}
+	}
+	return fmt.Errorf("cycle detected: %s", strings.Join(componentDetails, " -> "))
```

Comment:
To make tests more robust you can make a (unexported) custom error type that has the component list as a field and handles this formatting on the `Error` method. That way we can check the component list on tests instead of the error itself, and we won't have to update the tests in the future if we only update the error wording.

Reply:
I can appreciate what you're going for here but I'm not sure it's worth it given the tradeoffs.
- Validating the actual error message ensures we know it looks as expected.
- Updating the format is pretty easy. You can just run the tests, ensure the actual result looks correct, and if so replace the expected result with it.
- The error could use `[]graph.Node`, but then we have to type switch again. It's not difficult but seems like we end up duplicating most of this code in tests anyways, which seems more complex and less robust to me.

Reply:
Fair enough, those are good counterarguments :) We can leave it as is then and refactor later if the need arises